src/include/optimizer/optimizer_metadata.h

```diff
@@ -0,0 +1,256 @@
+
+  /**
+   * Gets the cost model
+   * @returns Cost Model
+   */
+  AbstractCostModel *GetCostModel() { return cost_model_; }
+
+  /**
+   * Gets the transaction
+   * @returns transaction
+   */
+  transaction::TransactionContext *GetTxn() { return txn_; }
+
+  /**
+   * Sets the transaction
+   * @param txn TransactionContext
+   */
+  void SetTxn(transaction::TransactionContext *txn) { txn_ = txn; }
+
+  /**
+   * Sets the CatalogAccessor
+   * @param accessor CatalogAccessor
+   */
+  void SetCatalogAccessor(catalog::CatalogAccessor *accessor) { accessor_ = accessor; }
+
+  /**
+   * Sets the StatsStorage
+   * @param storage StatsStorage
+   */
+  void SetStatsStorage(StatsStorage *storage) { stats_storage_ = storage; }
+
+  /**
+   * Set the task pool tracked by the OptimizerMetadata.
+   * Function passes ownership over task_pool
+   * @param task_pool Pointer to OptimizerTaskPool
+   */
+  void SetTaskPool(OptimizerTaskPool *task_pool) {
+    delete this->task_pool_;
+    this->task_pool_ = task_pool;
+  }
+
+  /**
+   * Converts an OperatorExpression into a GroupExpression.
+   * The GroupExpression is internal tracking that is focused on the concept
+   * of groups rather than operators like OperatorExpression.
+   *
+   * Subtrees of the OperatorExpression are individually converted to
+   * GroupExpressions and inserted into Memo, which allows for duplicate
+   * detection. The root GroupExpression, however, is not automatically
+   * inserted into Memo.
+   *
+   * @param expr OperatorExpression to convert
+   * @returns GroupExpression representing OperatorExpression
+   */
+  GroupExpression *MakeGroupExpression(common::ManagedPointer<OperatorExpression> expr) {
+    std::vector<GroupID> child_groups;
+    for (auto &child : expr->GetChildren()) {
+      // Create a GroupExpression for the child
+      auto gexpr = MakeGroupExpression(child);
+
+      // Insert into the memo (this allows for duplicate detection)
+      auto expr = memo_.InsertExpression(gexpr, false);
+      if (expr == nullptr) {
+        // Delete if need to (see InsertExpression spec)
+        child_groups.push_back(gexpr->GetGroupID());
+        delete gexpr;
+      } else {
+        child_groups.push_back(expr->GetGroupID());
+      }
+    }
+
+    return new GroupExpression(expr->GetOp(), std::move(child_groups));
+  }
+
+  /**
+   * Records a transformed OperatorExpression by creating a new group.
+   * A new group will be created if expr does not already exist.
+   *
+   * expr is not freed
+   *
+   * @param expr OperatorExpression to record
+   * @param gexpr Places the newly created GroupExpression
+   * @returns Whether the OperatorExpression already exists
+   */
+  bool RecordTransformedExpression(common::ManagedPointer<OperatorExpression> expr, GroupExpression **gexpr) {
+    return RecordTransformedExpression(expr, gexpr, UNDEFINED_GROUP);
+  }
+
+  /**
+   * Records a transformed OperatorExpression into a given group.
```

Comment:
Can you expand on what it means to "record" a transformation? Is this like a log for tracing or a queue for tasks?